xcm/src/v0/multiasset.rs

```diff
@@ -0,0 +1,383 @@
+		MultiAsset { fun, id: self }
+	}
+
+	/// Use the value of `self` along with a `fun` fungibility specifier to create the corresponding `WildMultiAsset`
+	/// wildcard (`AllOf`) value.
+	pub fn into_wild(self, fun: WildFungibility) -> WildMultiAsset {
+		WildMultiAsset::AllOf { fun, id: self }
+	}
+}
+
+/// Classification of whether an asset is fungible or not, along with a mandatory amount or instance.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Encode, Decode)]
+pub enum Fungibility {
+	Fungible(u128),
+	NonFungible(AssetInstance),
+}
+
+impl Fungibility {
+	pub fn is_kind(&self, w: WildFungibility) -> bool {
+		use {Fungibility::*, WildFungibility::{Fungible as WildFungible, NonFungible as WildNonFungible}};
+		matches!((self, w), (Fungible(_), WildFungible) | (NonFungible(_), WildNonFungible))
+	}
+}
+
+impl From<u128> for Fungibility {
+	fn from(amount: u128) -> Fungibility {
+		debug_assert_ne!(amount, 0);
+		Fungibility::Fungible(amount)
+	}
+}
+
+impl From<AssetInstance> for Fungibility {
+	fn from(instance: AssetInstance) -> Fungibility {
+		Fungibility::NonFungible(instance)
+	}
+}
+
+
+
+
+#[derive(Clone, Eq, PartialEq, Debug, Encode, Decode)]
+pub struct MultiAsset {
+	pub id: AssetId,
+	pub fun: Fungibility,
+}
+
+impl PartialOrd for MultiAsset {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl Ord for MultiAsset {
+	fn cmp(&self, other: &Self) -> Ordering {
+		match (&self.fun, &other.fun) {
+			(Fungibility::Fungible(..), Fungibility::NonFungible(..)) => Ordering::Less,
+			(Fungibility::NonFungible(..), Fungibility::Fungible(..)) => Ordering::Greater,
+			_ => (&self.id, &self.fun).cmp(&(&other.id, &other.fun)),
+		}
+	}
+}
+
+impl<A: Into<AssetId>, B: Into<Fungibility>> From<(A, B)> for MultiAsset {
+	fn from((id, fun): (A, B)) -> MultiAsset {
+		MultiAsset { fun: fun.into(), id: id.into() }
+	}
+}
+
+impl MultiAsset {
+	pub fn is_fungible(&self, maybe_id: Option<AssetId>) -> bool {
+		use Fungibility::*;
+		matches!(self.fun, Fungible(..)) && maybe_id.map_or(true, |i| i == self.id)
+	}
+
+	pub fn is_non_fungible(&self, maybe_id: Option<AssetId>) -> bool {
+		use Fungibility::*;
+		matches!(self.fun, NonFungible(..)) && maybe_id.map_or(true, |i| i == self.id)
+	}
+
+	/// Prepend a `MultiLocation` to a concrete asset, giving it a new root location.
+	pub fn reanchor(&mut self, prepend: &MultiLocation) -> Result<(), ()> {
+		self.id.reanchor(prepend)
+	}
+
+	/// Returns true if `self` is a super-set of the given `inner`.
+	pub fn contains(&self, inner: &MultiAsset) -> bool {
+		use Fungibility::*;
+		if self.id == inner.id {
+			match (&self.fun, &inner.fun) {
+				(Fungible(a), Fungible(i)) if a >= i => return true,
+				(NonFungible(a), NonFungible(i)) if a == i => return true,
+				_ => (),
+			}
+		}
+		false
+	}
+}
+
+
+
+/// A `Vec` of `MultiAsset`s. There may be no duplicate fungible items in here and when decoding, they must be sorted.
```

Comment:
Why not use a `BTreeSet` if you want to disallow duplication? (And it would also be sorted, right?)

Reply:
That's what `Assets` is for. This is meant to be lightweight.